File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/post/domain/Post.java
package io.twogether.nbe_5_7_2_02team.post.domain;

import io.twogether.nbe_5_7_2_02team.global.common.BaseEntity;
import io.twogether.nbe_5_7_2_02team.member.domain.Member;

import jakarta.persistence.*;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Post extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Lob
    @Column(nullable = false)
    private String content;

    @Setter
    @Enumerated(EnumType.STRING)
    private RecruitmentStatus recruitmentStatus;

    @Setter private LocalDate recruitmentDeadline;

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RecruitmentField> recruitmentFields = new ArrayList<>();

    @OneToMany(mappedBy = "post", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<PostTag> postTags = new ArrayList<>();

    @Setter
    @ElementCollection
    @CollectionTable(name = "post_image", joinColumns = @JoinColumn(name = "post_id"))
    @Column(name = "image_url")
    @OrderColumn(name = "image_order") // 이미지 순서
    private List<String> imageUrls = new ArrayList<>();

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    private Member member;

    @Builder
    public Post(
            String title,
            String content,
            RecruitmentStatus recruitmentStatus,
            List<String> imageUrls,
            Member member) {
        this.title = title;
        this.content = content;
        this.recruitmentStatus = recruitmentStatus;
        if (imageUrls != null) {
            this.imageUrls = imageUrls;
        }
        this.member = member;
    }

    public void update(String title, String content, RecruitmentStatus status) {
        if (title != null) this.title = title;
        if (content != null) this.content = content;
        if (status != null) this.recruitmentStatus = status;
    }
}

File: NBE_5_7_2_02TEAM/src/main/java/io/twogether/nbe_5_7_2_02team/post/domain/Post.kt
package io.twogether.nbe_5_7_2_02team.post.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import io.twogether.nbe_5_7_2_02team.global.common.BaseEntity
import io.twogether.nbe_5_7_2_02team.member.domain.Member
import io.twogether.nbe_5_7_2_02team.post.dto.request.PostUpdateRequest
import jakarta.persistence.*
import java.time.LocalDate

@Entity
class Post(
    @field:Column(nullable = false)
    var title: String,
    @field:Column(nullable = false, columnDefinition = "TEXT")
    @field:Lob
    var content: String,
    @field:Enumerated(EnumType.STRING)
    var recruitmentStatus: RecruitmentStatus,
    @field:JsonIgnore
    @field:ManyToOne(fetch = FetchType.LAZY)
    @field:JoinColumn(name = "member_id")
    val member: Member,
) : BaseEntity() {
    @field:Id
    @field:GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    @field:ElementCollection
    @field:CollectionTable(name = "post_image", joinColumns = [JoinColumn(name = "post_id")])
    @field:Column(name = "image_url")
    @field:OrderColumn(name = "image_order") // 이미지 순서
    var imageUrls: List<String> = listOf()

    var recruitmentDeadline: LocalDate? = null

    @field:OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    val recruitmentFields: MutableList<RecruitmentField> = arrayListOf()

    @field:OneToMany(mappedBy = "post", cascade = [CascadeType.ALL], orphanRemoval = true)
    val postTags: MutableList<PostTag> = arrayListOf()

    fun update(request: PostUpdateRequest) {
        title = request.title
        content = request.content
        recruitmentStatus = request.recruitmentStatus
        recruitmentDeadline = request.recruitmentDeadline
    }
}
